packages/flutter/lib/src/widgets/drag_boundary.dart

```diff
@@ -0,0 +1,91 @@
+///
+/// See also:
+///  * [DragBoundary], which uses this class.
+///
+/// `T` is a data class that defines the shape being dragged. For example, when dragging a rectangle within the boundary, `T` should be a `Rect`. 
+abstract class DragBoundaryDelegate<T> {
+  /// Returns whether the specified drag object position is within the boundary.
+  bool isWithinBoundary(T position);
+
+  /// Returns the position of the given drag object after moving it fully inside
+  /// the boundary with the shortest distance.
+  T nearestPositionWithinBoundary(T position);
+}
+
+class _DragBoundaryDelegateForRect extends DragBoundaryDelegate<Rect> {
+  _DragBoundaryDelegateForRect(this.boundary);
+  final Rect boundary;
+  @override
+  bool isWithinBoundary(Rect position) {
+    return boundary.contains(position.topLeft) && boundary.contains(position.bottomRight);
+  }
+
+  @override
+  Rect nearestPositionWithinBoundary(Rect position) {
+    assert(() {
+      if (boundary.right - position.width < boundary.left ||
+          boundary.bottom - position.height < boundary.top) {
+        throw FlutterError(
+          'The shape is larger than the boundary. '
+          'The shape width must be less than the boundary width, and the shape height must be less than the boundary height.',
+        );
+      }
+      return true;
+    }());
+    final double left = clampDouble(
+      position.left,
+      boundary.left,
+      boundary.right - position.width,
+    );
+    final double top = clampDouble(
+      position.top,
+      boundary.top,
+      boundary.bottom - position.height,
+    );
+    return Rect.fromLTWH(left, top, position.width, position.height);
+  }
+}
+
+/// Provides a [DragBoundaryDelegate] for its descendants whose bounds are those defined by this widget.
+///
+/// {@tool dartpad}
+/// This example demonstrates dragging a red box, constrained within the bounds
+/// of a green box.
+///
+/// ** See code in examples/api/lib/widgets/gesture_detector/gesture_detector.3.dart **
+/// {@end-tool}
+class DragBoundary extends InheritedWidget {
+  /// Creates a widget that provides a boundary to its descendants.
+  const DragBoundary({required super.child, super.key});
+
+  /// Retrieve the [DragBoundary] from the nearest ancestor to
+  /// get its [DragBoundaryDelegate] of [Rect].
+  ///
+  /// [useGlobalPosition] Specifies whether to use global position.
+  /// If false, the local position of the bounds are used.
+  static DragBoundaryDelegate<Rect>? forRectOf(BuildContext context, {bool useGlobalPosition = false}) {
+    final InheritedElement? element =
+        context.getElementForInheritedWidgetOfExactType<DragBoundary>();
+    if (element == null) {
+      return null;
+    }
+    final RenderBox? rb = element.findRenderObject() as RenderBox?;
+    assert(rb != null && rb.hasSize, 'DragBoundary is not available');
+    final Rect boundary = (useGlobalPosition ? rb!.localToGlobal(Offset.zero) : Offset.zero) & rb!.size;
```

Comment:
This will not work if `localToGlobal` contains scaling. You might want to use `Rect.fromPoints` with two `localToGlobals`.

Reply:
I'm not very familiar with scaling, do you mean like this? `Rect.fromPoints(rb!.localToGlobal(Offset.zero), rb.localToGlobal(rb.size.bottomRight(Offset.zero)));`

If you are still online, could you provide a commit suggestion for this? I will look into scaling later.

Reply:
> I'm not very familiar with scaling, do you mean like this? `Rect.fromPoints(rb!.localToGlobal(Offset.zero), rb.localToGlobal(rb.size.bottomRight(Offset.zero)));`

Yes, exactly.

However, when I tried to demonstrate (and verify) it by creating a demo, which uses `useGlobalPosition: true`, I find it so hard that I start to wonder how it's going to be useful at all. If we want the `GestureDetector` to be on the rectangle being dragged, and drag boundary must also wrap the rectangle, which means the rectangle's position must not be global, but at least relative to the drag boundary. On the other hand, if the rectangle is positioned on a global overlay, then the drag boundary is unlikely a parent of the rectangle. The only case I can think of is when the `GestureDetector` is as big as the drag boundary. My point is, can you create an example when `useGlobalPosition` is true?

Reply:
On the contrary, `useGlobalPosition: true` is actually more useful, as it is not always possible to obtain the local position of the drag object within the boundary during use. For example, if this PR is used for `ReorderableList`, during dragging, `ReorderableList` needs to use the given `DragBoundaryDelegate`, and here `useGlobalPosition: true` must be used.

However, if we can ensure that we are using the local coordinates of the boundary, `useGlobalPosition: false` can avoid the need for two coordinate conversions.